Unable to upload through IAM client

**Environment information**
* MWDB version (from `/about`): 2.8.0
* Installation method:
  - [ ] mwdb.cert.pl service
  - [ ] From PyPi (`pip install mwdb-core`)
  - [x] From docker-compose
  - [ ] Other (please explain)
* Plugins installed: A few, but they seem unrelated

**Behaviour the bug (what happened?)**

We have `iam_auth` set to `1`, so in `get_s3_client` utility we get a IAM boto3 client. Then we try to run `put_object` on it, but it doesn't work. This results in 500 from the backend.

**Expected behaviour**

Upload file successfully.

**Reproduction Steps**

set `iam_auth` to true in `mwdb.ini` or env variables and try to upload.

**Additional context**

```
2022-10-24 15:12:49[INFO] uWSGIWorker4Core0 - app.log_request:146 - request - path:/api/file - arguments:{} - method:POST - status:500 - response_time:0:00:00.386883 - response_size:37
2022-10-24 15:12:49 [ERROR] uWSGIWorker4Core0 - service.error_router:76 - Unhandled exception occurred
2022-10-24 15:12:49 Traceback (most recent call last):
2022-10-24 15:12:49 File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1516, in full_dispatch_request
2022-10-24 15:12:49 rv = self.dispatch_request()
2022-10-24 15:12:49 File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1502, in dispatch_request
2022-10-24 15:12:49 return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
2022-10-24 15:12:49 File "/usr/local/lib/python3.8/site-packages/flask_restful/__init__.py", line 467, in wrapper
2022-10-24 15:12:49 resp = resource(*args, **kwargs)
2022-10-24 15:12:49 File "/usr/local/lib/python3.8/site-packages/flask_limiter/extension.py", line 1036, in __inner
2022-10-24 15:12:49 return current_app.ensure_sync(obj)(*a, **k)
2022-10-24 15:12:49 File "/usr/local/lib/python3.8/site-packages/flask_limiter/extension.py", line 1036, in __inner
2022-10-24 15:12:49 return current_app.ensure_sync(obj)(*a, **k)
2022-10-24 15:12:49 File "/usr/local/lib/python3.8/site-packages/flask_limiter/extension.py", line 1036, in __inner
2022-10-24 15:12:49 return current_app.ensure_sync(obj)(*a, **k)
2022-10-24 15:12:49 [Previous line repeated 5 more times]
2022-10-24 15:12:49 File "/usr/local/lib/python3.8/site-packages/flask/views.py", line 84, in view
2022-10-24 15:12:49 return current_app.ensure_sync(self.dispatch_request)(*args, **kwargs)
2022-10-24 15:12:49 File "/usr/local/lib/python3.8/site-packages/flask_restful/__init__.py", line 582, in dispatch_request
2022-10-24 15:12:49 resp = meth(*args, **kwargs)
2022-10-24 15:12:49 File "/app/./mwdb/resources/__init__.py", line 47, in endpoint
2022-10-24 15:12:49 return f(*args, **kwargs)
2022-10-24 15:12:49 File "/app/./mwdb/resources/__init__.py", line 31, in endpoint
2022-10-24 15:12:49 return f(*args, **kwargs)
2022-10-24 15:12:49 File "/app/./mwdb/resources/file.py", line 187, in post
2022-10-24 15:12:49 return self.create_object(obj["options"])
2022-10-24 15:12:49 File "/app/./mwdb/resources/object.py", line 124, in create_object
2022-10-24 15:12:49 item, is_new = self._create_object(
2022-10-24 15:12:49 File "/app/./mwdb/resources/file.py", line 34, in _create_object
2022-10-24 15:12:49 return File.get_or_create(
2022-10-24 15:12:49 File "/app/./mwdb/model/file.py", line 130, in get_or_create
2022-10-24 15:12:49 get_s3_client(
2022-10-24 15:12:49 File "/usr/local/lib/python3.8/site-packages/botocore/client.py", line 861, in __getattr__
2022-10-24 15:12:49 raise AttributeError(
2022-10-24 15:12:49 AttributeError: 'IAM' object has no attribute 'put_object'
```

--- a/mwdb/core/util.py
+++ b/mwdb/core/util.py
@@ -17,6 +17,7 @@ import boto3
 import botocore.client
 import magic
 import ssdeep
+from botocore.credentials import InstanceMetadataFetcher, InstanceMetadataProvider
 from flask_restful import abort
 from flask_sqlalchemy import Pagination
 
@@ -176,8 +177,16 @@ def get_s3_client(
         else:
             endpoint_url = "http://" + endpoint
 
+    session_token = None
+
     if iam_auth:
-        return boto3.client("iam", endpoint_url=endpoint_url, region_name=region)
+        provider = InstanceMetadataProvider(
+            iam_role_fetcher=InstanceMetadataFetcher(timeout=1000, num_attempts=2)
+        )
+        creds = provider.load().get_frozen_credentials()
+        access_key = creds.access_key
+        secret_key = creds.secret_key
+        session_token = creds.token
 
     if access_key is None or secret_key is None:
         raise RuntimeError(
@@ -189,5 +198,6 @@ def get_s3_client(
         endpoint_url=endpoint_url,
         aws_access_key_id=access_key,
         aws_secret_access_key=secret_key,
+        aws_session_token=session_token,
         region_name=region,
     )

--- a/mwdb/core/util.py
+++ b/mwdb/core/util.py
@@ -17,7 +17,11 @@ import boto3
 import botocore.client
 import magic
 import ssdeep
-from botocore.credentials import InstanceMetadataFetcher, InstanceMetadataProvider
+from botocore.credentials import (
+    ContainerProvider,
+    InstanceMetadataFetcher,
+    InstanceMetadataProvider,
+)
 from flask_restful import abort
 from flask_sqlalchemy import Pagination
 
@@ -180,13 +184,20 @@ def get_s3_client(
     session_token = None
 
     if iam_auth:
-        provider = InstanceMetadataProvider(
-            iam_role_fetcher=InstanceMetadataFetcher(timeout=1000, num_attempts=2)
-        )
-        creds = provider.load().get_frozen_credentials()
-        access_key = creds.access_key
-        secret_key = creds.secret_key
-        session_token = creds.token
+        iam_providers = [
+            ContainerProvider(),
+            InstanceMetadataProvider(
+                iam_role_fetcher=InstanceMetadataFetcher(timeout=1000, num_attempts=2)
+            ),
+        ]
+
+        for provider in iam_providers:
+            creds = provider.load()
+            if creds:
+                access_key = creds.access_key
+                secret_key = creds.secret_key
+                session_token = creds.token
+                break
 
     if access_key is None or secret_key is None:
         raise RuntimeError(

